Issue #181 - Fixes failure of database functions if special characters used

yorkshire-pudding opened the pull request:
Fixes #181 - Database functions fail if db_password contains special characters
Used the `rawurlencode` and `rawurldecode` that core uses

I've tested using the lando recipe in the module and they all pass (details below).
<details>
PHPUnit 8.5.26 #StandWithUkraine

Bee Core
 ✔ Bee is installed correctly
 ✔ Unknown command triggers an error
 ✔ Missing required argument triggers an error
 ✔ Commands can be called via an alias
 ✔ Root global option works
 ✔ Yes global option works
 ✔ Debug global option works

Cache Commands
 ✔ Cache clear command works

Config Commands
 ✔ Config get command works
 ✔ Config set command works
 ✔ Config export command works
 ✔ Config import command works

Cron Commands
 ✔ Cron command works

DBCommands
 ✔ Db export command works
 ✔ Db import command works

DBLog Commands
 ✔ Log command works

Download Commands
 ✔ Download command works
 ✔ Download core command works

Help Commands
 ✔ Help command works

PHPCommands
 ✔ Eval command works

Projects Commands
 ✔ Projects command works
 ✔ Enable command works
 ✔ Disable command works
 ✔ Uninstall command works
 ✔ Theme default command works
 ✔ Theme admin command works

Update Commands
 ✔ Update db command works

User Commands
 ✔ Users command works
 ✔ User password command works
 ✔ User login command works

Time: 43.26 seconds, Memory: 10.00 MB

OK (30 tests, 89 assertions)
PHPUnit 8.5.26 #StandWithUkraine

Multisite
 ✔ Site global option works

Multisite Install Commands
 ✔ Install command works

Multisite Download Commands
 ✔ Download command works

Time: 15.84 seconds, Memory: 10.00 MB

OK (3 tests, 20 assertions)
</details>
What I couldn't quite figure out is how to work with the `.lando.yml` and `setup.sh` script that calls the `install.sh` to test with a password that contains special characters

Perhaps this needs some dedicated tests writing?

I'll try and test on my hosting where I can more easily set a complex password than in lando

yorkshire-pudding committed:
URL encode/decode of db user, pass and name

yorkshire-pudding committed:
URL encode/decode of db user, pass and name

yorkshire-pudding committed:
Remove echo line used for testing

yorkshire-pudding committed:
Remove echo line used for testing

yorkshire-pudding commented:
Tested on host.
Works with complex password in interactive mode but not in auto mode.  I think this is because the `parse_url` function doesn't work with characters that are invalid as part of a URL, whereas with interactive we encode each element.
I came up with a different idea initially but thought `parse_url` was more efficient.  I will revisit that previous idea as we want it to work with secure passwords in auto mode.
The other option is a breaking change and splitting the constituent parts of the db option into individual options so they can be encoded individually.

yorkshire-pudding committed:
Modify the function to accurately extract parts.

yorkshire-pudding committed:
Modify the function to accurately extract parts.

yorkshire-pudding commented:
Tested this with the lando recipe
<details>
PHPUnit 8.5.26 #StandWithUkraine

Bee Core
 ✔ Bee is installed correctly
 ✔ Unknown command triggers an error
 ✔ Missing required argument triggers an error
 ✔ Commands can be called via an alias
 ✔ Root global option works
 ✔ Yes global option works
 ✔ Debug global option works

Cache Commands
 ✔ Cache clear command works

Config Commands
 ✔ Config get command works
 ✔ Config set command works
 ✔ Config export command works
 ✔ Config import command works

Cron Commands
 ✔ Cron command works

DBCommands
 ✔ Db export command works
 ✔ Db import command works

DBLog Commands
 ✔ Log command works

Download Commands
 ✔ Download command works
 ✔ Download core command works

Help Commands
 ✔ Help command works

PHPCommands
 ✔ Eval command works

Projects Commands
 ✔ Projects command works
 ✔ Enable command works
 ✔ Disable command works
 ✔ Uninstall command works
 ✔ Theme default command works
 ✔ Theme admin command works

Update Commands
 ✔ Update db command works

User Commands
 ✔ Users command works
 ✔ User password command works
 ✔ User login command works

Time: 49.68 seconds, Memory: 10.00 MB

OK (30 tests, 89 assertions)
PHPUnit 8.5.26 #StandWithUkraine

Multisite
 ✔ Site global option works

Multisite Install Commands
 ✔ Install command works

Multisite Download Commands
 ✔ Download command works

Time: 16.71 seconds, Memory: 10.00 MB

OK (3 tests, 20 assertions)
</details>

Also tested on host and this now works with complex password both auto and interactively.
The only (unlikely) condition I can see tripping this up, and would be tricky to accommodate is if someone put a `:` character in their database username.

yorkshire-pudding committed:
Newline at end and enclose password with quote

yorkshire-pudding committed:
Newline at end and enclose password with quote

Commit:
Update Lando's PHP container to v7.4.

BWPanda commented on commands/db.bee.inc:
```diff
@@ -115,11 +115,28 @@ function db_import_bee_callback($arguments, $options) {
-  // Open SQL command-line.
-  $command = 'mysql --user=' . $db_info['username'] . ' --password=' . $db_info['password'] . ' --host=' . $db_info['host'] . ' ' . $db_info['database'];
-  proc_close(proc_open($command, array(STDIN, STDOUT, STDERR), $pipes));
+  // Define command line base
+  $command =  $mysql_function;
+  $command .= ' --user=' . rawurldecode($db_info['username']);
+  $command .= ' --password=\'' .  rawurldecode($db_info['password']) . '\'' ;
+  $command .= ' --host=' . $db_info['host'];
+  $command .= ' ' . rawurldecode($db_info['database']);
+
+  // Return the concatenated command
+  return $command;
```
```suggestion
  $options = '--user=' . rawurldecode($db_info['username']);
  $options .= ' --password=' . rawurldecode($db_info['password']);
  $options .= ' --host=' . $db_info['host'];
  $options .= ' ' . rawurldecode($db_info['database']);

  return $options;
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/db.bee.inc:
```diff
@@ -63,12 +63,12 @@ function db_export_bee_callback($arguments, $options) {
 
   // Export and compress the database.
-  $export_command = 'mysqldump --user=' . $db_info['username'] . ' --password=' . $db_info['password'] . ' --host=' . $db_info['host'] . ' ' . $db_info['database'] . ' | gzip > ' . $filename . '.gz';
+  $export_command = bee_compile_mysql_command('mysqldump') . ' | gzip > ' . $filename . '.gz';
```
```suggestion
  $export_command = 'mysqldump ' . db_bee_mysql_options($db_info) . ' | gzip > ' . $filename . '.gz';
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/db.bee.inc:
```diff
@@ -96,7 +96,7 @@ function db_import_bee_callback($arguments, $options) {
     $import_command .= "gunzip -c $filename | ";
   }
-  $import_command .= 'mysql --user=' . $db_info['username'] . ' --password=' . $db_info['password'] . ' --host=' . $db_info['host'] . ' ' . $db_info['database'];
+  $import_command .= bee_compile_mysql_command();
```
```suggestion
  $import_command .= 'mysql ' . db_bee_mysql_options($db_info);
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/db.bee.inc:
```diff
@@ -115,11 +115,28 @@ function db_import_bee_callback($arguments, $options) {
  * database credentials.
  */
 function sql_bee_callback($arguments, $options) {
+
+  // Open SQL command-line.
+  $command = bee_compile_mysql_command();
+  proc_close(proc_open($command, array(STDIN, STDOUT, STDERR), $pipes));
```
```suggestion
  // Get database info.
  $db_connection = Database::getConnectionInfo();
  $db_info = $db_connection['default'];
  
  // Open SQL command-line.
  $command = 'mysql ' . db_bee_mysql_options($db_info);
  proc_close(proc_open($command, array(STDIN, STDOUT, STDERR), $pipes));
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/install.bee.inc:
```diff
@@ -86,12 +86,19 @@ function install_bee_callback($arguments, $options) {
   // Get database connection string.
   if (!empty($options['db'])) {
-    $db = $options['db'];
+    // Split the passed URL into parts so can encode user, password and db name
+    $db_url = $options['db'];
+    $db_name = rawurlencode(bee_check_plain(bee_after_last('/', $db_url)));
+    $db_username = rawurlencode(bee_check_plain(bee_before(':', bee_after('mysql://', $db_url))));
+    $db_password = rawurlencode(bee_check_plain(bee_between($db_username . ':', '@' . bee_after_last('@', $db_url), $db_url)));
+    $db_host = bee_check_plain(bee_between_last('@', '/', $db_url));
+    // Re-compile the database URL using the sanitized parts
+    $db = "mysql://$db_username:$db_password@$db_host/$db_name";
```
Please update this whole file's changes to instead do something similar to what `install.sh` does: https://github.com/backdrop/backdrop/blob/8762dc9369529b6238dda03318394d0f46dc6640/core/scripts/install.sh#L124

Using `parse_url()` should avoid the need for all the helper functions, and `array_map('urldecode', ...)` should simplify things too.

yorkshire-pudding replied:
@BWPanda - I did try using this, but `parse_url()` is designed for standard hyperlinks and whenever I passed the connection string whole (`--auto` rather than interactive) it failed.  It will not work on a database connection URL with complex passwords.  Take for example `xYK1jxu3$RSOEmL#evqI`; it cannot parse it.  Both the `$` and the `#` cause problems and you can't quote them as that's not valid in a URL either.  I ultimately think `install.sh` is flawed in that it will only accept simple MySQL passwords, whereas MySQL will accept complex passwords and also the GUI installer will handle complex passwords because it takes each element individually.

See https://github.com/backdrop-contrib/bee/pull/183#issuecomment-1094071869

[See where it was initially included](https://github.com/backdrop-contrib/bee/pull/182/commits/e8f96881b3834408117a507064f112cff85c5694#diff-29f05d9cc78b79338561837d2f67cb48fc7cb56a644bed66ca8477208a66ed3e)

yorkshire-pudding replied:
@BWPanda - please can we discuss?

BWPanda replied:
Yeah, looks like core still suffers from the same problem: https://github.com/backdrop/backdrop-issues/issues/2904

I think the easiest solution then is to ditch the `db` option and just provide options for each component individually...

yorkshire-pudding replied:
Done - see comment below.

BWPanda reviewed (changes requested):
Thanks @yorkshire-pudding. It took me a while to get my head around all this. Mainly because I didn't notice at first the different places where you used `rawurlencode()` and `rawurldecode()` (I thought they were all decoding, and that threw me for a while). But finally worked it out and now left some comments/requested changes for you.

yorkshire-pudding committed:
Implements suggested changes to db.bee.inc

yorkshire-pudding committed:
Implements suggested changes to db.bee.inc

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -115,11 +115,28 @@ function db_import_bee_callback($arguments, $options) {
+ * Helper function to create the mysql command line
+ * using database credentials.
+ */
+function bee_compile_mysql_command($mysql_function = 'mysql') {
```
Done

yorkshire-pudding commented on commands/db.bee.inc:
```diff
@@ -115,11 +115,28 @@ function db_import_bee_callback($arguments, $options) {
+function bee_compile_mysql_command($mysql_function = 'mysql') {
   // Get database info.
   $db_connection = Database::getConnectionInfo();
   $db_info = $db_connection['default'];
```
Done

yorkshire-pudding committed:
Merge branch 'backdrop-contrib:1.x-1.x' into issue_181

yorkshire-pudding committed:
Split db settings into separate options

yorkshire-pudding committed:
Split db settings into separate options

yorkshire-pudding commented:
@BWPanda - As per our discussion, I've changed the code to split the database settings.  I've also changed the test (multi-site) where it uses this command.

I've run tests using the updated lando recipe:

<details>
PHPUnit 8.5.26 #StandWithUkraine

Bee Core
 ✔ Bee is installed correctly
 ✔ Unknown command triggers an error
 ✔ Missing required argument triggers an error
 ✔ Commands can be called via an alias
 ✔ Root global option works
 ✔ Yes global option works
 ✔ Debug global option works

Cache Commands
 ✔ Cache clear command works

Config Commands
 ✔ Config get command works
 ✔ Config set command works
 ✔ Config export command works
 ✔ Config import command works

Cron Commands
 ✔ Cron command works

DBCommands
 ✔ Db export command works
 ✔ Db import command works

DBLog Commands
 ✔ Log command works

Download Commands
 ✔ Download command works
 ✔ Download core command works

Help Commands
 ✔ Help command works

PHPCommands
 ✔ Eval command works

Projects Commands
 ✔ Projects command works
 ✔ Enable command works
 ✔ Disable command works
 ✔ Uninstall command works
 ✔ Theme default command works
 ✔ Theme admin command works

Update Commands
 ✔ Update db command works

User Commands
 ✔ Users command works
 ✔ User password command works
 ✔ User login command works

Time: 41.96 seconds, Memory: 10.00 MB

OK (30 tests, 89 assertions)
PHPUnit 8.5.26 #StandWithUkraine

Multisite
 ✔ Site global option works

Multisite Install Commands
 ✔ Install command works

Multisite Download Commands
 ✔ Download command works

Time: 15.69 seconds, Memory: 10.00 MB

OK (3 tests, 20 assertions)
</details>

yorkshire-pudding commented:
@BWPanda - please can you review the latest commit

BWPanda commented:
It's on my to do list.

BWPanda commented on commands/db.bee.inc:
```diff
@@ -120,6 +120,27 @@ function sql_bee_callback($arguments, $options) {
+ */
+function db_bee_mysql_options($db_info) {
+
+  $options =  $mysql_function;
```
This needs to go.

yorkshire-pudding replied:
Done

BWPanda commented on commands/db.bee.inc:
```diff
@@ -120,6 +120,27 @@ function sql_bee_callback($arguments, $options) {
+
+  $options =  $mysql_function;
+  $options .= ' --user=' . rawurldecode($db_info['username']);
+  $options .= ' --password=\'' .  rawurldecode($db_info['password']) . '\'' ;
```
Rather than having escaped characters, better to just use different quotes:

```suggestion
  $options .= " --password='" . rawurldecode($db_info['password']) . "'";
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/install.bee.inc:
```diff
@@ -59,8 +71,8 @@ function install_bee_command() {
         'bee install' => bt('Install Backdrop in interactive mode, providing information when prompted.'),
-        'bee install --db=mysql://admin:P@ssw0rd!@localhost/backdrop --auto' => bt('Install Backdrop automatically using the provided database credentials, and default settings for everything else.'),
-        'bee install --db=mysql://admin:P@ssw0rd!@localhost/backdrop --username=Root --password=N0tS3cur3 --email=root@mydomain.com --site-name="My awesome site!"' => bt('Install Backdrop using the given options, and be prompted for the rest.'),
+        'bee install --db-name=backdrop --db-user=admin --db-password=P@ssw0rd! --auto' => bt('Install Backdrop automatically using the provided database credentials, and default settings for everything else.'),
+        'bee install --db-name=backdrop --db-user=admin --db-password=P@ssw0rd! --username=Root --password=N0tS3cur3 --email=root@mydomain.com --site-name="My awesome site!"' => bt('Install Backdrop using the given options, and be prompted for the rest.'),
```
Neither of these examples mention the host parameter...

yorkshire-pudding replied:
Done

BWPanda commented on commands/install.bee.inc:
```diff
@@ -84,17 +96,40 @@ function install_bee_callback($arguments, $options) {
+  if (!empty($options['db-host'])) {
+    $db_host = $options['db-host'];
   }
   else {
-    $db_name = bee_check_plain(bee_input(bt('Database name'), '', TRUE));
-    $db_username = bee_check_plain(bee_input(bt('Database username'), '', TRUE));
-    $db_password = bee_check_plain(bee_input(bt('Database password'), '', TRUE));
     $db_host = bee_check_plain(bee_input(bt('Database host'), 'localhost', TRUE));
-    $db = "mysql://$db_username:$db_password@$db_host/$db_name";
   }
```
With the host, I think it'd be nice to just assume it's `localhost` when the user provides a db-name (i.e. they've provided info via the command (meaning they probably don't want to be prompted), but they haven't provided the host so it's likely just the default).

yorkshire-pudding replied:
Done

BWPanda reviewed (changes requested):
A few things to change please.

yorkshire-pudding committed:
Implement suggested changes

yorkshire-pudding committed:
Merge branch 'issue_181' of github.com:yorkshire-pudding/bee into issue_181

yorkshire-pudding commented:
Thanks @BWPanda for your feedback.  I've implemented these.

BWPanda commented on commands/db.bee.inc:
```diff
@@ -120,6 +120,26 @@ function sql_bee_callback($arguments, $options) {
+ *   A string of command options that can be appended to 'mysql' or 'mysqldump'.
+ */
+function db_bee_mysql_options($db_info) {
+
+  $options .= ' --user=' . rawurldecode($db_info['username']);
```
```suggestion
  $options = '--user=' . rawurldecode($db_info['username']);
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/install.bee.inc:
```diff
@@ -84,17 +96,41 @@ function install_bee_callback($arguments, $options) {
+  // Get database name.
+  if (!empty($options['db-name'])) {
+    $db_name = $options['db-name'];
+    $check_db_name_option = TRUE;
```
```suggestion
```

yorkshire-pudding replied:
Done.

BWPanda commented on commands/install.bee.inc:
```diff
@@ -84,17 +96,41 @@ function install_bee_callback($arguments, $options) {
+  if (!empty($options['db-host'])) {
+    $db_host = $options['db-host'];
+  }
+  else {
+    $check_db_name_option ? $db_host = 'localhost' : $db_host = bee_check_plain(bee_input(bt('Database host'), 'localhost', TRUE));
+  }
```
```suggestion
  elseif (!empty($options['db-name'])) {
    // If the user provided a DB name but not the host, we assume 'localhost'.
    $db_host = 'localhost';
  }
  else {
    $db_host = bee_check_plain(bee_input(bt('Database host'), 'localhost', TRUE));
  }
```

yorkshire-pudding replied:
Done.

yorkshire-pudding committed:
Implement suggested changes

yorkshire-pudding commented:
@BWPanda - Done those.  

Regarding checking the db name, are there any general principles about when you would test once and store in a variable versus checking the same test each time?  I'm learning plenty, but sometimes I can't see why one approach would be favoured over another and would like to understand more.

BWPanda commented:
@yorkshire-pudding I understand where you're coming from. If it were performing a DB query or reading from config files, then yes, doing that once and then storing the value in a local variable would be better. But in this case, checking the DB name is just checking a variable (albeit an array rather than a string), so there's no real benefit storing this value in a second variable for later. Hope that makes sense...?

yorkshire-pudding commented:
Thanks. That's a good explanation. Makes sense.

BWPanda commented on commands/db.bee.inc:
```diff
@@ -120,6 +120,26 @@ function sql_bee_callback($arguments, $options) {
+ *   A string of command options that can be appended to 'mysql' or 'mysqldump'.
+ */
+function db_bee_mysql_options($db_info) {
+
+  $options = ' --user=' . rawurldecode($db_info['username']);
```
Please remove this leading blank line and the leading space before the `--user` (that's already accounted for when we call `db_bee_mysql_options()` above).

```suggestion
  $options = '--user=' . rawurldecode($db_info['username']);
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/install.bee.inc:
```diff
@@ -13,9 +13,21 @@ function install_bee_command() {
-          'description' => bt("A database connection string of the form: 'mysql://user:pass@localhost/database_name'. See https://api.backdropcms.org/database-configuration for more details."),
-          'value' => bt('String'),
+        'db-name' => array(
+          'description' => bt("The database name."),
```
```suggestion
          'description' => bt("The name of the database to install into."),
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/install.bee.inc:
```diff
@@ -13,9 +13,21 @@ function install_bee_command() {
+          'value' => bt('Database name'),
+        ),
+        'db-user' => array(
+          'description' => bt("The username for the database."),
```
```suggestion
          'description' => bt("The username for connecting to the database."),
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/install.bee.inc:
```diff
@@ -13,9 +13,21 @@ function install_bee_command() {
+          'value' => bt('Database username'),
+        ),
+        'db-pass' => array(
+          'description' => bt("The password for the database. "),
```
```suggestion
          'description' => bt("The password for connecting to the database."),
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/install.bee.inc:
```diff
@@ -13,9 +13,21 @@ function install_bee_command() {
+          'value' => bt('Database password'),
+        ),
+        'db-host' => array(
+          'description' => bt("The host for the database relative to your webserver. Defaults to 'localhost'"),
```
I think we should mention the fact that specifying a DB name but not host causes this to default. Also not sure what "relative to your webserver" means ('relative' makes it  sound like a path)...

```suggestion
          'description' => bt("The host for the database. Defaults to 'localhost' when a database name is provided."),
```

yorkshire-pudding replied:
Done.

BWPanda commented on commands/install.bee.inc:
```diff
@@ -59,8 +71,8 @@ function install_bee_command() {
         'bee install' => bt('Install Backdrop in interactive mode, providing information when prompted.'),
-        'bee install --db=mysql://admin:P@ssw0rd!@localhost/backdrop --auto' => bt('Install Backdrop automatically using the provided database credentials, and default settings for everything else.'),
-        'bee install --db=mysql://admin:P@ssw0rd!@localhost/backdrop --username=Root --password=N0tS3cur3 --email=root@mydomain.com --site-name="My awesome site!"' => bt('Install Backdrop using the given options, and be prompted for the rest.'),
+        'bee install --db-name=backdrop --db-user=admin --db-password=P@ssw0rd! --db-host=db_server --auto' => bt('Install Backdrop automatically using the provided database credentials, and default settings for everything else.'),
+        'bee install --db-name=backdrop --db-user=admin --db-password=P@ssw0rd! --username=Root --password=N0tS3cur3 --email=root@mydomain.com --site-name="My awesome site!"' => bt('Install Backdrop using the given options, and be prompted for the rest.'),
```
I think the host in these examples should be switched, since the first examples gives the bare minimum and leaves the default values for the rest.

```suggestion
        'bee install --db-name=backdrop --db-user=admin --db-password=P@ssw0rd! --auto' => bt('Install Backdrop automatically using the provided database credentials, and default settings for everything else.'),
        'bee install --db-name=backdrop --db-user=admin --db-password=P@ssw0rd! --db-host=db_server --username=Root --password=N0tS3cur3 --email=root@mydomain.com --site-name="My awesome site!"' => bt('Install Backdrop using the given options, and be prompted for the rest.'),
```

yorkshire-pudding replied:
Done

BWPanda commented on commands/install.bee.inc:
```diff
@@ -84,17 +96,44 @@ function install_bee_callback($arguments, $options) {
-    $db = "mysql://$db_username:$db_password@$db_host/$db_name";
   }
+
+  // Compile database url
```
Comments should end with a full-stop.

```suggestion
  // Compile database URL.
```

yorkshire-pudding replied:
Done.

BWPanda reviewed (changes requested):
Sorry, looking more closely I found a few more things to change.

yorkshire-pudding committed:
Implement suggested changes

yorkshire-pudding commented:
Hi @BWPanda - I have done those.

BWpanda committed:
Update install.bee.inc
